hitl/loopback.py

```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/python3
+"""
+Author: Vertigo Designs, Ryan Summers
+
+Description: Loop-back integration tests for Stabilizer hardware
+"""
+import argparse
+import asyncio
+import json
+import sys
+
+from gmqtt import Client as MqttClient
+from miniconf import Miniconf
+
+# The minimum allowably loopback voltage error (difference between output set point and input
+# measured value).
+MINIMUM_VOLTAGE_ERROR = 0.010
+
+def _voltage_to_machine_units(voltage):
+    """ Convert a voltage to IIR machine units.
+
+    Args:
+        voltage: The voltage to convert
+
+    Returns:
+        The IIR machine-units associated with the voltage.
+    """
+    dac_range = 4.096 * 2.5
+    assert abs(voltage) <= dac_range, 'Voltage out-of-range'
+    return voltage / dac_range * 0x7FFF
+
+
+def static_iir_output(output_voltage):
+    """ Generate IIR configuration for a static output voltage.
+
+    Args:
+        output_voltage: The desired static IIR output voltage.
+
+    Returns
+        The IIR configuration to send over Miniconf.
+    """
+    machine_units = _voltage_to_machine_units(output_voltage)
+    return {
+        'y_min': machine_units,
+        'y_max': machine_units,
+        'y_offset': 0,
+        'ba': [1, 0, 0, 0, 0],
+    }
+
+
+class TelemetryReader:
+    """ Helper utility to read Stabilizer telemetry. """
+
+    @classmethod
+    async def create(cls, prefix, broker):
+        """Create a connection to the broker and an MQTT device using it."""
+        client = MqttClient(client_id='')
+        await client.connect(broker)
+        return cls(client, prefix)
+
+
+    def __init__(self, client, prefix):
+        """ Constructor. """
+        self.client = client
+        self._telemetry = []
+        self.client.on_message = self.handle_telemetry
+        self._telemetry_topic = f'{prefix}/telemetry'
+        self.client.subscribe(self._telemetry_topic)
+
+
+    def handle_telemetry(self, _client, topic, payload, _qos, _properties):
+        """ Handle incoming telemetry messages over MQTT. """
+        assert topic == self._telemetry_topic
+        self._telemetry.append(json.loads(payload))
```

Comment:
I'd pipe this through a [queue](https://docs.python.org/3/library/asyncio-queue.html) (`get_latest` becomes async as well).
Or even LiFoQueue as that seems to be what you are after here.
Or even, since you are only ever accessing the latest, simply only store the latest.